src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs

```diff
@@ -331,21 +331,21 @@ internal virtual NamedTypeSymbol LookupMetadataType(ref MetadataTypeName emitted
 
 Done:
             if (isTopLevel
-                && scope is not PENamespaceSymbol
                 && (emittedTypeName.ForcedArity == -1 || emittedTypeName.ForcedArity == emittedTypeName.InferredArity)
                 && GeneratedNameParser.TryParseFileTypeName(
                     emittedTypeName.UnmangledTypeName,
                     out string? displayFileName,
-                    out int ordinal,
+                    out byte[]? checksum,
                     out string? sourceName))
             {
                 // also do a lookup for file types from source.
                 namespaceOrTypeMembers = scope.GetTypeMembers(sourceName);
                 foreach (var named in namespaceOrTypeMembers)
                 {
-                    if (named.AssociatedSyntaxTree is SyntaxTree tree
-                        && getDisplayName(tree) == displayFileName
-                        && named.DeclaringCompilation.GetSyntaxTreeOrdinal(tree) == ordinal
+                    if (named.AssociatedFileIdentifier is FileIdentifier identifier
+                        && getDisplayName(identifier) == displayFileName
```

Comment:
>&& getDisplayName(identifier) == displayFileName [](http://example.com/codeflow?start=24&length=48)

Is the `getDisplayName()` comparison needed?

Reply:
Sorry, my earlier comment wasn't clear. Do we need to compare `displayFileName` at all? Isn't it sufficient to compare `FilePathChecksumOpt` for the file name?

Reply:
I think we should compare all components which contribute to the metadata name. The idea of GetTypeByMetadataName is for the given string to actually match the metadata name of the symbol.